ruby/ql/lib/codeql/ruby/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -748,51 +751,44 @@ module Trees {
     }
 
     final override predicate last(AstNode last, Completion c) {
+      super.last(last, c)
+      or
       last(this.getPattern(), last, c) and
       c.(MatchingCompletion).getValue() = false
       or
-      exists(BooleanCompletion bc, boolean flag, MatchingCompletion mc |
+      exists(BooleanCompletion bc, boolean flag |
+        bc.getValue() = flag.booleanNot() and
         lastCondition(last, bc, flag) and
         c =
           any(NestedMatchingCompletion nmc |
-            nmc.getInnerCompletion() = bc and nmc.getOuterCompletion() = mc
+            nmc.getInnerCompletion() = bc and nmc.getOuterCompletion().getValue() = false
           )
-      |
-        mc.getValue() = false and
-        bc.getValue() = flag.booleanNot()
-        or
-        not exists(this.getBody()) and
-        mc.getValue() = true and
-        bc.getValue() = flag
       )
-      or
-      last(this.getBody(), last, c)
-      or
-      not exists(this.getBody()) and
-      not exists(this.getCondition()) and
-      last(this.getPattern(), last, c)
     }
 
     final override predicate succ(AstNode pred, AstNode succ, Completion c) {
-      pred = this and
-      first(this.getPattern(), succ) and
-      c instanceof SimpleCompletion
-      or
-      exists(Expr next |
-        last(this.getPattern(), pred, c) and
-        c.(MatchingCompletion).getValue() = true and
-        first(next, succ)
-      |
-        next = this.getCondition()
+      last(this.getPattern(), pred, c) and
+      not c.(MatchingCompletion).getValue() = false and
```

Comment:
I think this should be `c.(MatchingCompletion).getValue() = true`.

Reply:
A pattern may also end with a simple completion, for example a `*var` never fails.

Reply:
Could write the following instead `c instanceof MatchingCompletion implies c.(MatchingCompletion).getValue() = true`